layabout.py

```diff
@@ -1,3 +1,287 @@
+import os
+import time
+import logging
+from typing import Any, DefaultDict, Dict, Callable, List, Tuple
+from inspect import signature
+from functools import wraps
+from collections import defaultdict
+
+from slackclient import SlackClient
+
+# TODO: This is a dependency of slackclient needed for exception handling. In
+# the future, we might be able to remove this. For now we don't want to declare
+# it as an explicit dependency.
+from websocket import WebSocketConnectionClosedException
+
 __author__ = 'Reilly Tucker Siemens'
 __email__ = 'reilly@tuckersiemens.com'
-__version__ = '0.1.0'
+__version__ = '1.0.0b'
+
+# Private type alias for the complex type of the handlers defaultdict.
+_Handlers = DefaultDict[str, List[Tuple[Callable, dict]]]
+
+log = logging.getLogger(__name__)
+
+
+class LayaboutError(Exception):
+    """ Base error for all Layabout exceptions. """
+
+
+class MissingSlackToken(LayaboutError):
+    """ Raised if a Slack API token could not be found. """
+
+
+class FailedConnection(LayaboutError, ConnectionError):
+    """
+    Raised if the Slack client could not connect to the Slack API.
+
+    Inherits from both :obj:`LayaboutError` and the built-in
+    :obj:`ConnectionError` for convenience in exception handling.
+    """
+
+
+class Layabout:
+    """
+    An event handler on top of the Slack RTM API.
+
+    Args:
+        name: A unique name for this :obj:`Layabout` instance.
+        env_var: The environment variable to try to load a Slack API token
+            from.
+
+    Attributes:
+        name (str): A unique name for this :obj:`Layabout` instance.
+        slack (SlackClient): A :obj:`slackclient.SlackClient` instance.
+            Initially :obj:`None`, the attribute is set after calling the
+            :obj:`Layabout.connect` or :obj:`Layabout.run`
+            methods.
+
+    Example:
+
+        .. code-block:: python
+
+           from layabout import Layabout
+
+           layabout = Layabout('app')
+
+
+           @layabout.handle('message')
+           def echo(slack, event):
+               \"\"\" Echo all messages seen by the app. \"\"\"
+               channel = event['channel']
+               message = event['text']
+               subtype = event.get('subtype')
+
+               # Avoid an infinite loop of echoing our own messages.
+               if subtype != 'bot_message':
+                   slack.rtm_send_message(channel, message)
+
+           layabout.run()
+    """
+    def __init__(self, name: str, env_var: str = 'SLACK_API_TOKEN') -> None:
+        # TODO: Consider keyword only arguments.
```

Comment:
No, I mean keyword-only arguments from [PEP 3102](https://www.python.org/dev/peps/pep-3102/) to enforce that users can only call these methods with keyword arguments rather than positional arguments, thus ensuring a greater level of backwards compatibility if I chose to do something like change the order of arguments.

In this case that might look something like this (note the `*` in the signature):
```python
class Layabout:
    def __init__(self, *, name: str, env_var: str = 'SLACK_API_TOKEN') -> None:
        ...
```
then callers would encounter a `TypeError` unless they specifically used keyword arguments.
```python
>>> layabout = Layabout()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: __init__() missing 1 required keyword-only argument: 'name'
>>> layabout = Layabout('app')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: __init__() takes 1 positional argument but 2 were given
>>> layabout = Layabout(name='app')
```
Later, I could then decide that I wanted `name` to be optional and `env_var` to come first in the signature
```python
class Layabout:
    def __init__(self, *, env_var: str = 'SLACK_API_TOKEN', name: str = None) -> None:
```
and it wouldn't matter at all since all users would have been prohibited from writing code that would break backwards compatibility here.

Reply:
I'd totally forgotten about this feature.

I think I personally lean on the side of keyword only arguments being best used for optional arguments. I think it's less clear that name is a required argument from the signature.